subprojects/core-api/src/main/java/org/gradle/api/provider/ListProperty.java

```diff
@@ -18,16 +18,47 @@
 
 import org.gradle.api.Incubating;
 
+import javax.annotation.Nullable;
+import java.util.Collection;
 import java.util.List;
 
 /**
  * Represents a property whose type is a {@link List} of elements of type {@link T}.
  *
- * <p><b>Note:</b> This interface is not intended for implementation by build script or plugin authors. An instance of this class can be created through the factory method {@link org.gradle.api.provider.ProviderFactory#listProperty(Class)}.
+ * <p><b>Note:</b> This interface is not intended for implementation by build script or plugin authors. An instance of
+ * this class can be created through the factory method {@link org.gradle.api.provider.ProviderFactory#listProperty(Class)}.
  *
  * @param <T> the type of elements.
  * @since 4.3
  */
 @Incubating
 public interface ListProperty<T> extends PropertyState<List<T>> {
+    /**
+     * Adds an element to the property without evaluating existing providers already present.
+     *
+     * @param element The element, can be null.
+     * @since 4.4
+     */
+    void add(T element);
+
+    /**
+     * Adds an element to the property given by the provider without evaluating existing providers already present.
+     * set. This property will track the value of the provider and query its value each time the value of the property
+     * is queried. When the provider has no value, the element in the property will also have no value.
+     *
+     * @param provider Provider
+     * @since 4.4
+     */
+    void add(Provider<? extends T> provider);
```

Comment:
What happens to the existing value of this property? What happens if it had no value?

Reply:
I discussed with @big-guy and we settled on:
- Existing values are kept as it only adds a value to the list
- If the property as no value, adding will succeed but it will throw an `IllegalStateException` when querying the base provider as it has no values.
-  If the provider has no value, it will throw an `IllegalStateException` only when querying the provider.